### PR TITLE
Added file_change_recipients command for pasta command line

### PR DESCRIPTION
**Pull request by ShubhamPandey28**
I have added a command using the relations between the files changed and recipients for patches to give all the possible recipients for a patch. 
[sample.patch](https://pastebin.com/HfGeN3mP)  used. 

Working proof of this command
```
(lfd-pasta) dark_matter@DarkMatter ~/l/l/PaStA (stream-3)> ./pasta file_change_recipients --gen --dump CSV --patch sample.patch --out outfile.lst
2020-08-02 20:43:10,525 PaStA           INFO     Cmdline: ./pasta file_change_recipients --gen --dump CSV --patch sample.patch --out outfile.lst
2020-08-02 20:43:10,526 pypasta.Config  INFO     Loading configuration: linux
2020-08-02 20:43:10,612 Repository.Mbox INFO     Loading mailbox subsystem
2020-08-02 20:43:10,617 Repository.Mbox INFO       ↪ loaded invalid mail index: found 4669 invalid mails
2020-08-02 20:43:10,617 Repository.Mbox INFO     Loading public inboxes
2020-08-02 20:43:13,897 Repository.Mbox INFO       ↪ loaded mail index for alsa-devel@alsa-project.org (shard 0): found 213500 mails
2020-08-02 20:43:14,500 ange_recipients INFO     Evaluating recipients realtion with respect to files changed
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4186/4186 [00:12<00:00, 327.15it/s]
2020-08-02 20:43:27,551 ange_recipients INFO     The related recipient list for the patch is stored in /home/dark_matter/linux_work/lfd-pasta/PaStA/outfile.lst
2020-08-02 20:43:27,654 PaStA           INFO     Shutting down
(lfd-pasta) dark_matter@DarkMatter ~/l/l/PaStA (stream-3)> cat outfile.lst
andriy.shevchenko@linux.intel.com
pavan.k.s@intel.com
cezary.rojewski@intel.com
pierre-louis.bossart@linux.intel.com
liam.r.girdwood@linux.intel.com
lgirdwood@gmail.com
linux-kernel@vger.kernel.org
amit.kucheria@linaro.org
broonie@kernel.org
linux-pm@vger.kernel.org
yang.jie@linux.intel.com
alsa-devel@alsa-project.org
mateusz.gorski@linux.intel.com
kuninori.morimoto.gx@renesas.com
s@alsa-project.org
tiwai@suse.com⏎ 
(lfd-pasta) dark_matter@DarkMatter ~/l/l/PaStA (stream-3)> ./pasta file_change_recipients --use CSV --patch sample.patch --out outfile.lst
2020-08-02 20:43:31,590 PaStA           INFO     Cmdline: ./pasta file_change_recipients --use CSV --patch sample.patch --out outfile.lst
2020-08-02 20:43:31,590 pypasta.Config  INFO     Loading configuration: linux
2020-08-02 20:43:31,676 Repository.Mbox INFO     Loading mailbox subsystem
2020-08-02 20:43:31,682 Repository.Mbox INFO       ↪ loaded invalid mail index: found 4669 invalid mails
2020-08-02 20:43:31,682 Repository.Mbox INFO     Loading public inboxes
2020-08-02 20:43:35,009 Repository.Mbox INFO       ↪ loaded mail index for alsa-devel@alsa-project.org (shard 0): found 213500 mails
2020-08-02 20:43:35,626 ange_recipients INFO     Loading recipients from /home/dark_matter/linux_work/lfd-pasta/PaStA/CSV
2020-08-02 20:43:35,649 ange_recipients INFO     The related recipient list for the patch is stored in /home/dark_matter/linux_work/lfd-pasta/PaStA/outfile.lst
2020-08-02 20:43:35,716 PaStA           INFO     Shutting down

```

This command still does not consider the following fact,
for a patch touching file A and B,
it is sent to recipient X because it touches file A and to recipient Y because it touches B.

I that case we need to study for P(file touched by diff | recipient X receives patch).
In conclusion, I think bayesian model might work fine in this case. But I am not sure about this idea and may be it needs to be disproved.

Suggested-by: @rralf  @bulwahn 
Signed-off-by: Shubhamkumar Pandey <b18194@students.iitmandi.ac.in>

### CODE REVIEW

**Comment by bulwahn** [comment]
> This command still does not consider the following fact,
> for a patch touching file A and B,
> it is sent to recipient X because it touches file A and to recipient Y because it touches B.
> 
> I that case we need to study for P(file touched by diff | recipient X receives patch).

What are you talking about? That is exactly what you are computing; of course, one could argue that weights should be distributed for every patch with some strategy.

> In conclusion, I think bayesian model might work fine in this case. But I am not sure about this idea and may be it needs to be disproved.
> 
> I was not sure that I have the authority to include someone in authors of the file, so I am tagging the developers with which I had discussion according to linux kernel code of conduct.
> 

What does the kernel code of conduct have to do with this? Please provide a pointer to what you are referring. I do not believe you know what is stated in the Linux kernel code of conduct. Please stop using words you do not know and understand.


Lukas

**Comment by ShubhamPandey28** [comment]
> > This command still does not consider the following fact,
> > for a patch touching file A and B,
> > it is sent to recipient X because it touches file A and to recipient Y because it touches B.
> > I that case we need to study for P(file touched by diff | recipient X receives patch).
> 
> What are you talking about? That is exactly what you are computing; of course, one could argue that weights should be distributed for every patch with some strategy.

I think that here the thing is calculated is P(recipient | file touched X) ,RECIPIENT[X]recipient]/total_patches_X stood for that. but RECIPIENT[X][recipient] /total_patches_recipient is P(file X | recipient). It's not big a deal actually its just dividing the same variable by different counts. but both are different as far as probability is concerned. Of course one knows

> 
> > In conclusion, I think bayesian model might work fine in this case. But I am not sure about this idea and may be it needs to be disproved.
> > I was not sure that I have the authority to include someone in authors of the file, so I am tagging the developers with which I had discussion according to linux kernel code of conduct.
> 
> What does the kernel code of conduct have to do with this? Please provide a pointer to what you are referring. I do not believe you know what is stated in the Linux kernel code of conduct. Please stop using words you do not know and understand.

Thanks, I will keep that in mind.
Sorry I was refering to patch tags (Acked-by, Suggested-by, Tested-by, Reviewed-by, Reported-by, Fixes) used in linux kernel patches. Seriously, I didn't think that this might be of any trouble here.

@bulwahn  I have updated the PR message.

Thanks,
-Shubhamkumar Pandey
> 
> Lukas

File: bin/pasta_file_change_recipients.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+from logging import getLogger
+
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
+from pypasta import *
+from pypasta.LinuxMailCharacteristics import email_get_recipients
+from pypasta.Repository.Patch import Diff
+
+log = getLogger(__name__[-15:])
+
+RECIPIENTS = None
```
Comment (rralf) [inline bin/pasta_file_change_recipients.py]: Better pass this variable around than making use of a global variable.

File: bin/pasta_file_change_recipients.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            for recipient in recipients:
+                RECIPIENTS[filename][recipient] += 1
+    
+    if f_recipients:
+        with open(f_recipients, 'w') as f:
+            f.write('filename, recipients, count\n')
+            for filename, ctr in RECIPIENTS.items():
+                ctr = sorted(ctr.items(), key = lambda x: x[1], reverse=True)
+                f.write(f"{filename}, {','.join(map(lambda x: x[0]+','+str(x[1]),ctr))}\n")
+                RECIPIENTS[filename] = set(map(lambda x: x[0], ctr))
```
Comment (rralf) [inline bin/pasta_file_change_recipients.py]: What happens in this line? I don't get it.

File: bin/pasta_file_change_recipients.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        for filename in patch.diff.affected:
+            for recipient in recipients:
+                RECIPIENTS[filename][recipient] += 1
+    
+    if f_recipients:
+        with open(f_recipients, 'w') as f:
+            f.write('filename, recipients, count\n')
+            for filename, ctr in RECIPIENTS.items():
+                ctr = sorted(ctr.items(), key = lambda x: x[1], reverse=True)
+                f.write(f"{filename}, {','.join(map(lambda x: x[0]+','+str(x[1]),ctr))}\n")
```
Comment (rralf) [inline bin/pasta_file_change_recipients.py]: Uh, black magic! Why don't you simply use python-pickle to dump RECIPIENTS?

File: bin/pasta_file_change_recipients.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    elif args.f_recipients:
+        args.f_recipients = os.path.join(os.getcwd(), args.f_recipients)
+        log.info(f"Loading recipients from {args.f_recipients}")
+        RECIPIENTS = defaultdict()
+        with open(args.f_recipients, 'r') as f:
+            lines = f.readlines()[1:]
+            for line in lines:
+                line = line.strip().split(",")
+                RECIPIENTS[line[0]] = set([ line[i].strip() for i in range(len(line)) if i%2 ])
```
Comment (rralf) [inline bin/pasta_file_change_recipients.py]: I bet we have a bug here. I saw some strange results... Better use python-pickle.

File: bin/pasta_file_change_recipients.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    args = parser.parse_args(argv)
+    repo = config.repo
+    repo.register_mbox(config)
+    global RECIPIENTS
+
+    if args._generate:
+        log.info('Evaluating recipients realtion with respect to files changed')
+        if args.f_recipients:
+            args.f_recipients = os.path.join(os.getcwd(), args.f_recipients)
+        dump_csv(config, args.f_recipients)
```
Comment (rralf) [inline bin/pasta_file_change_recipients.py]: Better approach: Try to load the pickle-file with the recipients, and check if anything is missing. So if the file doesn't exist, it will be generated, if it exists, it will be loaded, and missing data can be appended.

**Review by rralf — CHANGES_REQUESTED** [review]
Before leaving more comments - is this actually what I should have a look at? Thanks!

File: bin/pasta_file_change_recipients.py (hunk 1 above)
Comment (ShubhamPandey28) [reply bin/pasta_file_change_recipients.py]: Okay

File: bin/pasta_file_change_recipients.py (hunk 2 above)
Comment (ShubhamPandey28) [reply bin/pasta_file_change_recipients.py]: Just writing to file in `filename,mail_id,count,mail_id,count...` format
and taking the RECIPIENT[filename] = set of all mail_ids.

File: bin/pasta_file_change_recipients.py (hunk 3 above)
Comment (ShubhamPandey28) [reply bin/pasta_file_change_recipients.py]: @rralf  I asked you before for that for introducing a cache parameter in the config object. I saw the how pkl files are handled in the system and thought to introduce f_recipient_pkl in config object, but it seemed to be wrong to touch Config. So do I make a pkl file inside resource directory along with like patch-group and other caches without making an attribute inside Config?

File: bin/pasta_file_change_recipients.py (hunk 4 above)
Comment (ShubhamPandey28) [reply bin/pasta_file_change_recipients.py]: Sure

File: bin/pasta_file_change_recipients.py (hunk 5 above)
Comment (ShubhamPandey28) [reply bin/pasta_file_change_recipients.py]: Sure

**Comment by ShubhamPandey28** [comment]
I mistakenly force pushed this commit out of my branch. So this patch got terminated. I will shortly continue with a new one.

**Comment by ShubhamPandey28** [comment]
I have opened it again as https://github.com/lfd/PaStA/pull/74

File: bin/pasta_file_change_recipients.py (hunk 3 above)
Comment (rralf) [reply bin/pasta_file_change_recipients.py]: Simply choose a constant filename for the moment. We can later introduce a proper name when it comes to integration.

Comment (ShubhamPandey28) [reply bin/pasta_file_change_recipients.py]: I did, I also thought that would be better for now. So I did it that way in https://github.com/lfd/PaStA/pull/74, placed it with other pkl files in resource directory.
